File: idea/src/org/jetbrains/jet/plugin/quickfix/MakeOverriddenMemberOpenFix.java
/*
 * Copyright 2010-2013 JetBrains s.r.o.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.jetbrains.jet.plugin.quickfix;

import com.google.common.collect.Sets;
import com.intellij.codeInsight.intention.IntentionAction;
import com.intellij.openapi.editor.Editor;
import com.intellij.openapi.project.Project;
import com.intellij.openapi.util.Condition;
import com.intellij.psi.PsiElement;
import com.intellij.psi.PsiFile;
import com.intellij.util.IncorrectOperationException;
import com.intellij.util.containers.ContainerUtil;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import org.jetbrains.jet.lang.descriptors.CallableMemberDescriptor;
import org.jetbrains.jet.lang.descriptors.DeclarationDescriptor;
import org.jetbrains.jet.lang.diagnostics.Diagnostic;
import org.jetbrains.jet.lang.psi.JetDeclaration;
import org.jetbrains.jet.lang.psi.JetFile;
import org.jetbrains.jet.plugin.JetBundle;
import org.jetbrains.jet.plugin.caches.resolve.ResolvePackage;
import org.jetbrains.jet.plugin.project.ResolveSessionForBodies;

import java.util.*;

import static org.jetbrains.jet.lang.descriptors.CallableMemberDescriptor.Kind.*;
import static org.jetbrains.jet.lang.resolve.DescriptorToSourceUtils.descriptorToDeclaration;
import static org.jetbrains.jet.lexer.JetTokens.OPEN_KEYWORD;

public class MakeOverriddenMemberOpenFix extends JetIntentionAction<JetDeclaration> {
    private final List<PsiElement> overriddenNonOverridableMembers = new ArrayList<PsiElement>();
    private final List<String> containingDeclarationsNames = new ArrayList<String>();

    public MakeOverriddenMemberOpenFix(@NotNull JetDeclaration declaration) {
        super(declaration);
    }

    @Override
    public boolean isAvailable(@NotNull Project project, Editor editor, PsiFile file) {
        if (!super.isAvailable(project, editor, file) || !(file instanceof JetFile)) {
            return false;
        }

        // When running single test 'isAvailable()' is invoked multiple times, so we need to clear lists.
        overriddenNonOverridableMembers.clear();
        containingDeclarationsNames.clear();

        ResolveSessionForBodies resolveSession = ResolvePackage.getLazyResolveSession((JetFile) file);
        DeclarationDescriptor descriptor = resolveSession.resolveToDescriptor(element);
        if (!(descriptor instanceof CallableMemberDescriptor)) return false;

        for (CallableMemberDescriptor overriddenDescriptor : getAllDeclaredNonOverridableOverriddenDescriptors(
                (CallableMemberDescriptor) descriptor)) {
            assert overriddenDescriptor.getKind() == DECLARATION : "Can only be applied to declarations.";
            PsiElement overriddenMember = descriptorToDeclaration(overriddenDescriptor);
            if (overriddenMember == null || !QuickFixUtil.canModifyElement(overriddenMember)) {
                return false;
            }
            String containingDeclarationName = overriddenDescriptor.getContainingDeclaration().getName().asString();
            overriddenNonOverridableMembers.add(overriddenMember);
            containingDeclarationsNames.add(containingDeclarationName);
        }
        return overriddenNonOverridableMembers.size() > 0;
    }

    @NotNull
    private static Collection<CallableMemberDescriptor> getAllDeclaredNonOverridableOverriddenDescriptors(
            @NotNull CallableMemberDescriptor callableMemberDescriptor
    ) {
        Set<CallableMemberDescriptor> result = Sets.newHashSet();
        Collection<CallableMemberDescriptor>
                nonOverridableOverriddenDescriptors = retainNonOverridableMembers(callableMemberDescriptor.getOverriddenDescriptors());
        for (CallableMemberDescriptor overriddenDescriptor : nonOverridableOverriddenDescriptors) {
            CallableMemberDescriptor.Kind kind = overriddenDescriptor.getKind();
            if (kind == DECLARATION) {
                result.add(overriddenDescriptor);
            }
            else if (kind == FAKE_OVERRIDE || kind == DELEGATION) {
                result.addAll(getAllDeclaredNonOverridableOverriddenDescriptors(overriddenDescriptor));
            }
            else if (kind == SYNTHESIZED) {
                // do nothing, final synthesized members can't be made open
            }
            else {
                throw new UnsupportedOperationException("Unexpected callable kind " + kind);
            }
        }
        return result;
    }

    @NotNull
    private static Collection<CallableMemberDescriptor> retainNonOverridableMembers(
            @NotNull Collection<? extends CallableMemberDescriptor> callableMemberDescriptors
    ) {
        return ContainerUtil.filter(callableMemberDescriptors, new Condition<CallableMemberDescriptor>() {
            @Override
            public boolean value(CallableMemberDescriptor descriptor) {
                return !descriptor.getModality().isOverridable();
            }
        });
    }

    @NotNull
    @Override
    public String getText() {
        if (overriddenNonOverridableMembers.size() == 1) {
            return JetBundle.message("make.element.modifier", containingDeclarationsNames.get(0) + "." + element.getName(), OPEN_KEYWORD);
        }

        StringBuilder declarations = new StringBuilder();
        Collections.sort(containingDeclarationsNames);
        for (int i = 0; i < containingDeclarationsNames.size() - 2; i++) {
            declarations.append(containingDeclarationsNames.get(i));
            declarations.append(", ");
        }
        declarations.append(containingDeclarationsNames.get(containingDeclarationsNames.size() - 2));
        declarations.append(" and ");
        declarations.append(containingDeclarationsNames.get(containingDeclarationsNames.size() - 1));
        return JetBundle.message("make.element.in.classifiers.open", element.getName(), declarations.toString());
    }

    @NotNull
    @Override
    public String getFamilyName() {
        return JetBundle.message("add.modifier.family");
    }

    @Override
    public void invoke(@NotNull Project project, Editor editor, JetFile file) throws IncorrectOperationException {
        for (PsiElement overriddenMember : overriddenNonOverridableMembers) {
            overriddenMember.replace(AddModifierFix.addModifierWithDefaultReplacement(overriddenMember, OPEN_KEYWORD, project, false));
        }
    }

    @NotNull
    public static JetSingleIntentionActionFactory createFactory() {
        return new JetSingleIntentionActionFactory() {
            @Nullable
            @Override
            public IntentionAction createAction(Diagnostic diagnostic) {
                JetDeclaration declaration = QuickFixUtil.getParentElementOfType(diagnostic, JetDeclaration.class);
                assert declaration != null;
                return new MakeOverriddenMemberOpenFix(declaration);
            }
        };
    }
}

File: idea/src/org/jetbrains/kotlin/idea/quickfix/MakeOverriddenMemberOpenFix.kt
/*
 * Copyright 2010-2019 JetBrains s.r.o. and Kotlin Programming Language contributors.
 * Use of this source code is governed by the Apache 2.0 license that can be found in the license/LICENSE.txt file.
 */

package org.jetbrains.kotlin.idea.quickfix

import com.intellij.codeInsight.intention.IntentionAction
import com.intellij.openapi.editor.Editor
import com.intellij.openapi.project.Project
import com.intellij.psi.SmartPsiElementPointer
import com.intellij.psi.util.CachedValueProvider
import com.intellij.psi.util.PsiModificationTracker
import org.jetbrains.kotlin.descriptors.CallableMemberDescriptor
import org.jetbrains.kotlin.descriptors.CallableMemberDescriptor.Kind.*
import org.jetbrains.kotlin.descriptors.MemberDescriptor
import org.jetbrains.kotlin.descriptors.Modality
import org.jetbrains.kotlin.descriptors.isOverridable
import org.jetbrains.kotlin.diagnostics.Diagnostic
import org.jetbrains.kotlin.idea.KotlinBundle
import org.jetbrains.kotlin.idea.caches.resolve.resolveToDescriptorIfAny
import org.jetbrains.kotlin.idea.core.util.CachedValue
import org.jetbrains.kotlin.idea.core.util.getValue
import org.jetbrains.kotlin.idea.refactoring.canRefactor
import org.jetbrains.kotlin.idea.util.actualsForExpected
import org.jetbrains.kotlin.idea.util.isExpectDeclaration
import org.jetbrains.kotlin.lexer.KtTokens.OPEN_KEYWORD
import org.jetbrains.kotlin.psi.KtCallableDeclaration
import org.jetbrains.kotlin.psi.KtDeclaration
import org.jetbrains.kotlin.psi.KtFile
import org.jetbrains.kotlin.psi.psiUtil.createSmartPointer
import org.jetbrains.kotlin.psi.psiUtil.getNonStrictParentOfType
import org.jetbrains.kotlin.resolve.DescriptorToSourceUtils
import org.jetbrains.kotlin.resolve.lazy.BodyResolveMode

private typealias DeclarationPointer = SmartPsiElementPointer<KtCallableDeclaration>

class MakeOverriddenMemberOpenFix(declaration: KtDeclaration) : KotlinQuickFixAction<KtDeclaration>(declaration) {

    private val myQuickFixInfo: QuickFixInfo by CachedValue(declaration.project) {
        CachedValueProvider.Result.createSingleDependency(computeInfo(), PsiModificationTracker.MODIFICATION_COUNT)
    }

    private val containingDeclarationsNames
        get() = myQuickFixInfo.declNames

    private val overriddenNonOverridableMembers
        get() = myQuickFixInfo.declarations

    private fun computeInfo(): QuickFixInfo {
        val element = element ?: return QUICKFIX_UNAVAILABLE
        val overriddenNonOverridableMembers = mutableListOf<DeclarationPointer>()
        val containingDeclarationsNames = mutableListOf<String>()
        val descriptor = element.resolveToDescriptorIfAny(BodyResolveMode.FULL) as? CallableMemberDescriptor ?: return QUICKFIX_UNAVAILABLE

        for (overriddenDescriptor in getAllDeclaredNonOverridableOverriddenDescriptors(descriptor)) {
            assert(overriddenDescriptor.kind == DECLARATION) { "Can only be applied to declarations." }
            val overriddenMember = DescriptorToSourceUtils.descriptorToDeclaration(overriddenDescriptor)
            if (overriddenMember == null || !overriddenMember.canRefactor() || overriddenMember !is KtCallableDeclaration ||
                overriddenMember.modifierList?.hasModifier(OPEN_KEYWORD) == true
            ) {
                return QUICKFIX_UNAVAILABLE
            }

            overriddenDescriptor.takeIf { overriddenMember.isExpectDeclaration() }?.actualsForExpected()?.forEach {
                if (it is MemberDescriptor && it.modality < Modality.OPEN) {
                    val member = DescriptorToSourceUtils.descriptorToDeclaration(it)
                    if (member == null || !member.canRefactor() || member !is KtCallableDeclaration) {
                        return QUICKFIX_UNAVAILABLE
                    }
                    overriddenNonOverridableMembers.add(member.createSmartPointer())
                }
            }

            val containingDeclarationName = overriddenDescriptor.containingDeclaration.name.asString()
            overriddenNonOverridableMembers.add(overriddenMember.createSmartPointer())
            containingDeclarationsNames.add(containingDeclarationName)
        }
        return QuickFixInfo(overriddenNonOverridableMembers, containingDeclarationsNames)
    }

    override fun isAvailable(project: Project, editor: Editor?, file: KtFile): Boolean {
        return overriddenNonOverridableMembers.isNotEmpty()
    }

    override fun getText(): String {
        val element = element ?: return ""
        if (containingDeclarationsNames.size == 1) {
            val name = containingDeclarationsNames[0] + "." + element.name
            return KotlinBundle.message("make.0", "$name $OPEN_KEYWORD")
        }
        val sortedDeclarationNames = containingDeclarationsNames.sorted()
        val declarations = sortedDeclarationNames.subList(0, sortedDeclarationNames.size - 1).joinToString(", ") +
                " ${KotlinBundle.message("configuration.text.and")} " +
                sortedDeclarationNames.last()
        return KotlinBundle.message("make.0.in.1.open", element.name.toString(), declarations)
    }

    override fun getFamilyName(): String = KotlinBundle.message("add.modifier")

    override fun invoke(project: Project, editor: Editor?, file: KtFile) {
        for (overriddenMember in overriddenNonOverridableMembers) {
            overriddenMember.element?.addModifier(OPEN_KEYWORD)
        }
    }

    companion object : KotlinSingleIntentionActionFactory() {
        private data class QuickFixInfo(val declarations: List<DeclarationPointer>, val declNames: List<String>)

        private val QUICKFIX_UNAVAILABLE = QuickFixInfo(emptyList(), emptyList())

        private fun getAllDeclaredNonOverridableOverriddenDescriptors(
            callableMemberDescriptor: CallableMemberDescriptor
        ): Collection<CallableMemberDescriptor> {
            val result = hashSetOf<CallableMemberDescriptor>()
            val nonOverridableOverriddenDescriptors = retainNonOverridableMembers(callableMemberDescriptor.overriddenDescriptors)
            for (overriddenDescriptor in nonOverridableOverriddenDescriptors) {
                when (overriddenDescriptor.kind) {
                    DECLARATION ->
                        result.add(overriddenDescriptor)

                    FAKE_OVERRIDE, DELEGATION ->
                        result.addAll(getAllDeclaredNonOverridableOverriddenDescriptors(overriddenDescriptor))

                    SYNTHESIZED -> {
                    } /* do nothing */

                    else -> throw UnsupportedOperationException("Unexpected callable kind ${overriddenDescriptor.kind}")
                }
            }
            return result
        }

        private fun retainNonOverridableMembers(
            callableMemberDescriptors: Collection<CallableMemberDescriptor>
        ): Collection<CallableMemberDescriptor> {
            return callableMemberDescriptors.filter { !it.isOverridable }
        }

        override fun createAction(diagnostic: Diagnostic): IntentionAction? {
            val declaration = diagnostic.psiElement.getNonStrictParentOfType<KtDeclaration>()!!
            return MakeOverriddenMemberOpenFix(declaration)
        }
    }
}
